### PR TITLE
fix: in memory cache for coroutines

### PR DESCRIPTION
**Pull request by floredelasteyrie**

### CODE REVIEW

File: test/test_cache.py [hunk 1]
```diff
@@ -120,31 +103,66 @@ def tick(self, value: int = test_ttl):
...
-        backend.cache.__setitem__.assert_called_with(key, value)
-        backend.cache.reset_mock()
+        co_cache = self.sample_cacheable_function(value)
+        assert co_cache.done is False
+        assert backend.get(key)[0] == co_cache
 
-        backend.get(key)
-        backend.cache.get.assert_called_with(key, NO_VALUE)
+        co_cache = self.sample_cacheable_function(value)
+        awaited = await co_cache
```
Comment (d-tw) [inline test/test_cache.py:158]: Maybe assert that we have the same co_cache object

File: test/test_cache.py [hunk 2]
```diff
@@ -120,31 +103,66 @@ def tick(self, value: int = test_ttl):
         self.time += value + 1
```
Comment (d-tw) [inline test/test_cache.py:103]: We need to test the persistence in isolation maybe using `pyfakefs`?

We also need to test the co-routine caching - I don't think we have a test for that yet

**Review by d-tw — CHANGES_REQUESTED** [review]

File: make/ci.Makefile [hunk 3]
```diff
@@ -89,7 +89,7 @@ test: clean test-unit test-integration coverage ## Run all tests
 # i.e. SKIP_COVERAGE=1 when the `test` target has been called, or SKIP_COVERAGE=0 otherwise
 define run-pytest
 	@SKIP_COVERAGE=$(ALL_TESTS) $(MAKE) clean-coverage
-	poetry run coverage run --context=$(COVERAGE_CONTEXT) -m pytest -vv --maxfail=1 --ff ./test
+	poetry run coverage run --context=$(COVERAGE_CONTEXT) -m pytest -vv -s --maxfail=1 --ff ./test
```
Comment (d-tw) [inline make/ci.Makefile:92]:
```suggestion
	poetry run coverage run --context=$(COVERAGE_CONTEXT) -m pytest -vv --maxfail=1 --ff ./test
```

File: src/outcome/utils/cache.py [hunk 4]
```diff
@@ -140,16 +157,32 @@ def __init__(self, arguments):
...
         self.cache[key] = value
         if self.persisted_cache_path:
             self.persist_cache()
 
     def delete(self, key):
-        self.cache.pop(key)
-        if self.persisted_cache_path:
+        self.coroutine_cache.pop(key, None)
+
+        if self.cache.pop(key, None) is not None and self.persisted_cache_path:
```
Comment (d-tw) [inline src/outcome/utils/cache.py:185]:
```suggestion
        
        sentinel = object()
        if self.cache.pop(key, sentinel) is not sentinel and self.persisted_cache_path:
```

**Review by d-tw — APPROVED** [review]

File: src/outcome/utils/cache.py (hunk 4 above)
Comment (floredelasteyrie) [reply src/outcome/utils/cache.py:185]: Is it to make the difference is a `None` value in `cache` ?

Comment (d-tw) [reply src/outcome/utils/cache.py:185]: Yep, creating an instance of `object`, then checking its identity with `is not` ensures that we can distinguish an absent value from any other value

**Pull request merged**